rule_builder.py

```diff
@@ -0,0 +1,573 @@
+import dataclasses
+import itertools
+import operator
+from collections import defaultdict
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Any, ClassVar, Self
+
+from worlds.AutoWorld import World
+
+if TYPE_CHECKING:
+    from BaseClasses import CollectionState, Entrance, Item, MultiWorld
+    from NetUtils import JSONMessagePart
+    from Options import CommonOptions, Option
+
+OPERATORS = {
+    "eq": operator.eq,
+    "ne": operator.ne,
+    "gt": operator.gt,
+    "lt": operator.lt,
+    "ge": operator.ge,
+    "le": operator.le,
+    "contains": operator.contains,
+}
+
+
+@dataclasses.dataclass()
+class Rule:
+    """Base class for a static rule used to generate a"""
+
+    options: dict[str, Any] = dataclasses.field(default_factory=dict, kw_only=True)
+    """A mapping of option_name to value to restrict what options are required for this rule to be active.
+    An operator can be specified with a double underscore and the operator after the option name, eg `opt__le`
+    """
+
+    def _passes_options(self, options: "CommonOptions") -> bool:
+        """Tests if the given world options pass the requirements for this rule"""
+        for key, value in self.options:
+            parts = key.split("__", maxsplit=1)
+            option_name = parts[0]
+            operator = parts[1] if len(parts) > 1 else "eq"
+            opt: Option[Any] = getattr(options, option_name)
+            if not OPERATORS[operator](opt.value, value):
+                return False
+        return True
+
+    def _instantiate(self, world: "RuleWorldMixin") -> "Resolved":
+        """Create a new resolved rule for this world"""
+        return self.Resolved(player=world.player)
+
+    def resolve(self, world: "RuleWorldMixin") -> "Resolved":
+        """Resolve a rule with the given world"""
+        if not self._passes_options(world.options):
+            return False_.Resolved(player=world.player)
+
+        instance = self._instantiate(world)
+        rule_hash = hash(instance)
+        if rule_hash not in world.rule_ids:
+            world.rule_ids[rule_hash] = instance
+        return world.rule_ids[rule_hash]
+
+    def to_json(self) -> Any:
+        """Returns a JSON-serializable definition of this rule"""
+        return {
+            "rule": self.__class__.__name__,
+            "args": {field.name: getattr(self, field.name, None) for field in dataclasses.fields(self)},
+        }
+
+    @classmethod
+    def from_json(cls, data: Any) -> Self:
+        if not isinstance(data, Mapping):
+            raise ValueError("Invalid data format for parsed json")
```

Comment:
If it needs to be a `Mapping` in order for the function to work, then why not put that information in the `data` parameter type annotation?

Reply:
Also does this even work? Mapping is a protocol type so e.g. Mapping is not actually in dict's class hierarchy

Reply:
I'm not 100% on how I want to go with this. For the base implementation, a mapping is required and enforced at runtime. However, a user may create a custom rule that defines its own `to_json` and `from_json` that may or may not use a mapping, which is why I left that type as `Any`.

re BM: yes, you can use `issubclass` and `isinstance` on these ABCs: https://docs.python.org/3/library/collections.abc.html

Reply:
Decided to change this to have `Mapping` in the signature. If it's a problem in the future we can deal with it then